src/fstbin/fsts-concat.cc

```diff
@@ -0,0 +1,111 @@
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+
+#include "base/kaldi-common.h"
+#include "util/common-utils.h"
+#include "fstext/fstext-utils.h"
+#include "fstext/kaldi-fst-io.h"
+
+
+int main(int argc, char *argv[]) {
+  try {
+    using namespace kaldi;
+    using namespace fst;
+    typedef kaldi::int32 int32;
+    typedef kaldi::uint64 uint64;
+
+    const char *usage =
+        "Reads kaldi archives with FSTs. Concatenates the fsts from all the rspecifiers.\n"
+        "The fsts to concatenate must have same key. The sequencing is given by the position of arguments.\n"
+        "\n"
+        "Usage: fsts-concat [options] <fsts-rspecifier1> <fsts-rspecifier2> ... <fsts-wspecifier>\n"
+        " e.g.: fsts-concat scp:fsts1.scp scp:fsts2.scp ... ark:fsts_out.ark\n"
+        "\n"
+        "see also: fstconcat (from the OpenFst toolkit)\n";
+
+    ParseOptions po(usage);
+
+    po.Read(argc, argv);
+
+    if (po.NumArgs() < 3) {
+      po.PrintUsage();
+      exit(1);
+    }
+
+    std::string fsts_rspecifier = po.GetArg(1),
+        fsts_wspecifier = po.GetArg(po.NumArgs());
+
+    SequentialTableReader<VectorFstHolder> fst_reader(fsts_rspecifier);
+    std::vector<RandomAccessTableReader<VectorFstHolder>*> fst_readers;
+    TableWriter<VectorFstHolder> fst_writer(fsts_wspecifier);
+
+    for (int32 i=2; i<po.NumArgs(); i++)
+      fst_readers.push_back(new RandomAccessTableReader<VectorFstHolder>(po.GetArg(i)));
+
+    int32 n_done = 0,
+          n_skipped = 0;
+
+    for (; !fst_reader.Done(); fst_reader.Next()) {
+      std::string key = fst_reader.Key();
+
+      // check that the key exists in all 'fst_readers',
+      bool skip_key = false;
+      for (int32 i=0; i<fst_readers.size(); i++) {
+        if (!fst_readers[i]->HasKey(key)) {
+          KALDI_WARN << "Skippng '" << key << "'"
+            << " due to missing the fst in " << i+2 << ". <rspecifier> : "
+            << "'" << po.GetArg(i+2) << "'";
+          skip_key=true;
+        }
+      }
+      if (skip_key) {
+        n_skipped++;
+        continue;
+      }
+
+      // Concatenate!
+      VectorFst<StdArc> fst_out = fst_readers.back()->Value(key);
+      // loop from (last-1) to first, as 'prepending' the fsts is faster,
+      // see: http://www.openfst.org/twiki/bin/view/FST/ConcatDoc
+      for (int32 i=fst_readers.size()-2; i>=0; i--) {
+        fst::Concat(fst_readers[i]->Value(key), &fst_out);
+      }
+      // finally, prepend the fst from the 'Sequential' reader,
+      fst::Concat(fst_reader.Value(), &fst_out);
+
+      // write the output,
+      fst_writer.Write(key, fst_out);
+      n_done++;
+    }
+
+    // cleanup,
```

Comment:
please start comments with capitals and end with period, not comma.  It's google style guide... sorry to be picky.